runtime/compiler/ilgen/Walker.cpp

```diff
@@ -6419,6 +6444,152 @@ TR_J9ByteCodeIlGenerator::genFlattenableWithFieldWithHelper(uint16_t fieldCpInde
    push(helperCallNode);
    }
 
+static TR::SymbolReference * createLoadFieldSymRef(TR::Compilation * comp, TR_OpaqueClassBlock * fieldClass, const char * fieldname)
+   {
+   const TR::TypeLayout *fieldClassLayout = comp->typeLayout(fieldClass);
+   size_t fieldClassFieldCount = fieldClassLayout->count();
+
+   for (size_t idx = 0; idx < fieldClassFieldCount; idx++)
+      {
+      const TR::TypeLayoutEntry &fieldEntry = fieldClassLayout->entry(idx);
+      if (!strcmp(fieldname, fieldEntry._fieldname))
+         {
+         auto * fieldSymRef = comp->getSymRefTab()->findOrFabricateShadowSymbol(fieldClass,
+                                                                              fieldEntry._datatype,
+                                                                              fieldEntry._offset,
+                                                                              fieldEntry._isVolatile,
+                                                                              fieldEntry._isPrivate,
+                                                                              fieldEntry._isFinal,
+                                                                              fieldEntry._fieldname,
+                                                                              fieldEntry._typeSignature
+                                                                              );
+         return fieldSymRef;
+         }
+      }
+
+   TR_ASSERT_FATAL(false, "Did not find the matching fieldname %s", fieldname);
+   return NULL;
+   }
+
+void
+TR_J9ByteCodeIlGenerator::genFlattenableWithField(uint16_t fieldCpIndex, TR_OpaqueClassBlock * valueClass)
+   {
+   /* An example on what the tree with flattened fields would look like
+    *
+    * value Point2D {
+    *    public final int x;
+    *    public final int y;
+    * }
+    *
+    * value FlattenedLine2D {
+    *    public final Point2D st;
+    *    public final Point2D en;
+    *
+    *    public static FlattenedLine2D withSt(FlattenedLine2D line, Point2D st) {
+    *       0: aload_1
+    *       1: aload_0
+    *       3: withfield #3 // Field st:QPoint2D;
+    *       6: astore_2
+    *       7: aload_2
+    *       8: areturn
+    *    }
+    * }
+    *
+    * method="FlattenedLine2D.withSt(QFlattenedLine2D;QPoint2D;)QFlattenedLine2D;"
+    * 3: JBwithfield
+    * /--- trees inserted ------------------------
+    * n7n      (  0)  NULLCHK on n3n [#32]
+    * n6n      (  2)    iloadi  Point2D.x I[#355  final Point2D.x I +4]
+    * n3n      (  2)      aload  <parm 1 F>[#353  Parm]
+    * n9n      (  0)  NULLCHK on n3n [#32]
```

Comment:
this NULLCHK is redundant - could we avoid generating it if it is still being generated?

Reply:
`genFlattenableWithField()` uses `loadInstance()` to load each field. When it gets to `loadInstance`, it doesn't know if the nullchk has been added to the address because of the last load. To avoid the redundancy, initially I was thinking of adding an argument to `loadInstance()` to avoid the redundant nullchk, something like `loadInstance(TR::SymbolReference * symRef, bool skipNullChk = false) `and in `genFlattenableWithField()`, we track when to pass` skipNullChk=true` to `loadInstance()`.  But this way, the API gets more complicated. Not sure if that's the route we'd prefer.

Reply:
ok I guess we can leave it for now, but you should make sure the redundant checks are eliminated successfully by the optimizer. I agree the boolean in the API is not ideal

Reply:
It looks to me the redundant nullchk is eliminated after optimization.

```
n1n       BBStart <block_2> (freq 8001)                                                      
n54n        GlRegDeps                                                                        
n55n          aRegLoad esi   <parm 1 F>[#354  Parm] [flags 0x40000107 0x0 ] (X>=0 SeenRealReference sharedMemory ) 
n56n          aRegLoad eax   <parm 0 Q>[#353  Parm] [flags 0x40000107 0x0 ] (X>=0 SeenRealReference sharedMemory ) 
n7n       NULLCHK on n55n [#32]                                                              
n6n         iloadi  Point2D.x I[#356  final Point2D.x I +4] [flags 0x20603 0x200 ] (cannotOverflow ) 
n55n          ==>aRegLoad
n11n      NULLCHK on n56n [#32]                                                              
n10n        iloadi  FlattenedLine2D.en.x I[#358  final FlattenedLine2D.en.x I +12] [flags 0x20603 0x200 ] (cannotOverflow ) 
n56n          ==>aRegLoad
n15n      treetop                                                                            
n14n        newvalue  jitNewValue[#100  helper Method] [flags 0x400 0x0 ] (Identityless sharedMemory ) 
n5n           loadaddr  FlattenedLine2D[#355  Static] [flags 0x18307 0x0 ]                   
n6n           ==>iloadi
n8n           iloadi  Point2D.y I[#357  final Point2D.y I +8] [flags 0x20603 0x200 ] (cannotOverflow ) 
n55n            ==>aRegLoad
n10n          ==>iloadi
n12n          iloadi  FlattenedLine2D.en.y I[#359  final FlattenedLine2D.en.y I +16] [flags 0x20603 0x200 ] (cannotOverflow ) 
n56n            ==>aRegLoad
n18n      areturn                                                                            
n14n        ==>newvalue
n2n       BBEnd </block_2> =====      
```